File: ocaml/deriv.ml
open Printf

type expr =
  | Int of int
  | Var of string
  | Add of expr * expr
  | Mul of expr * expr
  | Pow of expr * expr
  | Ln of expr

let rec pown a = function
  | 0 -> 1
  | 1 -> a
  | n ->
    let b = pown a (n / 2) in
    b * b * (if n mod 2 = 0 then 1 else a)

let rec add = function
  | Int m, Int n -> Int(m + n)
  | Int 0, f | f, Int 0 -> f
  | f, Int n -> add(Int n, f)
  | f, Add(Int n, g) -> add(Int n, add(f, g))
  | Add(f, g), h -> add(f, add(g, h))
  | f, g -> Add(f, g)
let rec mul = function
  | Int m, Int n -> Int(m * n)
  | Int 0, f | f, Int 0 -> Int 0
  | Int 1, f | f, Int 1 -> f
  | f, Int n -> mul(Int n, f)
  | f, Mul(Int n, g) -> mul(Int n, mul(f, g))
  | Mul(f, g), h -> mul(f, mul(g, h))
  | f, g -> Mul(f, g)
let rec pow = function
  | Int m, Int n -> Int(pown m n)
  | f, Int 0 -> Int 1
  | f, Int 1 -> f
  | Int 0, f -> Int 0
  | f, g -> Pow(f, g)
let ln = function
  | Int 1 -> Int 0
  | f -> Ln f

let rec d x = function
  | Var y when x=y -> Int 1
  | Int _ | Var _ -> Int 0
  | Add(f, g) -> add(d x f, d x g)
  | Mul(f, g) -> add(mul(f, d x g), mul(g, d x f))
  | Pow(f, g) -> mul(pow(f, g), add(mul(mul(g, d x f), pow(f, Int(-1))), mul(ln f, d x g)))
  | Ln f -> mul(d x f, pow(f, Int(-1)))

let rec count = function
  | Int _ | Var _ -> 1
  | Add(f, g) | Mul(f, g) | Pow(f, g) -> count f + count g
  | Ln f -> count f

let rec string_of () = function
  | Int n -> sprintf "%d" n
  | Var x -> x
  | Add(f, g) -> sprintf "%a + %a" string_of f string_of g
  | Mul(f, g) -> sprintf "%a*%a" (bracket 2) f (bracket 2) g
  | Pow(f, g) -> sprintf "%a^%a" (bracket 2) f (bracket 3) g
  | Ln f -> sprintf "ln(%a)" string_of f
and prec = function
  | Int _ | Var _ | Ln _ -> 4
  | Pow _ -> 3
  | Mul _ -> 2
  | Add _ -> 1
and bracket outer () expr =
  if outer > prec expr then
    sprintf "(%a)" string_of expr
  else
    string_of () expr

let string_of_expr expr =
  let n = count expr in
  if n > 100 then sprintf "<<%d>>" n else
    sprintf "%a" string_of expr

let rec nest n f x =
  if n = 0 then x else
    nest (n-1) f (f x)

let deriv f =
  let d = d "x" f in
  printf "D(%s) = %s\n%!" (string_of_expr f) (string_of_expr d);
  d

let () =
  let x = Var "x" in
  let f = pow(x, x) in
  ignore(nest (int_of_string Sys.argv.(1)) deriv f)

File: swift/deriv.swift
enum Expr {
  case Int(n: Int)
  case Var(x: String)
  indirect case Add(f: Expr, g: Expr)
  indirect case Mul(f: Expr, g: Expr)
  indirect case Pow(f: Expr, g: Expr)
  indirect case Ln(f: Expr)
}

func pown(a: Int, b: Int) -> Int {
  switch (b) {
    case 0 : return 1
    case 1 : return a
    case let n :
      let b = pown(a: a, b: n / 2)
      return b * b * (n % 2 == 0 ? 1 : a)
  }
}

func add(f: Expr, g: Expr) -> Expr {
  switch (f, g) {
    case (let .Int(n: m), let .Int(n: n)) : return .Int(n: m + n)
    case (.Int(n: 0), let f) : return f
    case (let f, .Int(n: 0)) : return f
    case (let f, let .Int(n)) : return add(f: .Int(n: n), g: f)
    case (let f, let .Add(.Int(n), g)) : return add(f: .Int(n: n), g: add(f: f, g: g))
    case (let .Add(f, g), let h) : return add(f: f, g: add(f: g, g: h))
    case (let f, let g) : return .Add(f: f, g: g)
  }
}

func mul(f: Expr, g: Expr) -> Expr {
  switch (f, g) {
    case (let .Int(n: m), let .Int(n: n)) : return .Int(n: m * n)
    case (.Int(n: 0), _) : return .Int(n: 0)
    case (_, .Int(n: 0)) : return .Int(n: 0)
    case (.Int(n: 1), let f) : return f
    case (let f, .Int(n: 1)) : return f
    case (let f, let .Int(n: n)) : return mul(f: .Int(n: n), g: f)
    case (let f, let .Mul(.Int(n: n), g)) : return mul(f: .Int(n: n), g: mul(f: f, g: g))
    case (let .Mul(f: f, g: g), let h) : return mul(f: f, g: mul(f: g, g: h))
    case (let f, let g) : return .Mul(f: f, g: g)
  }
}

func pow(f: Expr, g: Expr) -> Expr {
  switch (f, g) {
    case (let .Int(n: m), let .Int(n: n)) : return .Int(n: pown(a: m, b: n))
    case (_, .Int(n: 0)) : return .Int(n: 1)
    case (let f, .Int(n: 1)) : return f
    case (.Int(n: 0), _) : return .Int(n: 0)
    case (let f, let g) : return .Pow(f: f, g: g)
  }
}

func ln(f: Expr) -> Expr {
  switch (f) {
    case .Int(n: 1) : return .Int(n: 0)
    case let f : return .Ln(f: f)
  }
}

func d(x: String, f: Expr) -> Expr {
  switch (f) {
    case .Int(n: _) : return .Int(n: 0)
    case let .Var(x: y) : if x == y { return .Int(n: 1) } else { return .Int(n: 0) }
    case let .Add(f: f, g: g) : return add(f: d(x: x, f: f), g: d(x: x, f: g))
    case let .Mul(f: f, g: g) : return add(f: mul(f: f, g: d(x: x, f: g)), g: mul(f: g, g: d(x: x, f: f)))
    case let .Pow(f: f, g: g) : return mul(f: pow(f: f, g: g), g: add(f: mul(f: mul(f: g, g: d(x: x, f: f)), g: pow(f: f, g: .Int(n: -1))), g: mul(f: ln(f: f), g: d(x: x, f: g))))
    case let .Ln(f: f) : return mul(f: d(x: x, f: f), g: pow(f: f, g: .Int(n: -1)))
  }
}

func count(f: Expr) -> Int {
  switch (f) {
    case .Int(n: _) : return 1
    case .Var(x: _) : return 1
    case let .Add(f: f, g: g) : return count(f: f) + count(f: g)
    case let .Mul(f: f, g: g) : return count(f: f) + count(f: g)
    case let .Pow(f: f, g: g) : return count(f: f) + count(f: g)
    case let .Ln(f: f) : return count(f: f)
  }
}

func stringOfExpr(f: Expr) -> String {
  switch (f) {
    case let .Int(n: n) : return String(n)
    case let .Var(x: x) : return x
    case let .Add(f: f, g: g) : return "(" + stringOfExpr(f: f) + " + " + stringOfExpr(f: g) + ")"
    case let .Mul(f: f, g: g) : return "(" + stringOfExpr(f: f) + " * " + stringOfExpr(f: g) + ")"
    case let .Pow(f: f, g: g) : return "(" + stringOfExpr(f: f) + "^" + stringOfExpr(f: g) + ")"
    case let .Ln(f: f) : return "ln(" + stringOfExpr(f: f) + ")"
  }
}

func stringOf(f: Expr) -> String {
  let n = count(f: f)
  if n > 100 {
    return "<<" + String(n) + ">>"
  } else {
    return stringOfExpr(f: f)
  }
}

let x = Expr.Var(x: "x")

let f = pow(f: x, g: x)

func nest(n: Int, f: ((Expr) -> Expr), x: Expr) -> Expr {
  if n == 0 { return x } else {
    return nest(n: n-1, f: f, x: f(x))
  }
}

var dx = { (f: Expr) -> Expr in
  var df = d(x: "x", f: f)
  print("D(" + stringOf(f: f) + ") = " + stringOf(f: df))
  return df
}

var n = Int(CommandLine.arguments[1])

print(count(f: nest(n: n!, f: dx, x: f)))
